Add option for generating diffs in HTML format.

vadz opened the pull request:
Generate text/html emails with inline colorized diffs if "htmldiffs" option is
set.

This is based on the original patch from Marc Mengel (see PR #74), but was
almost completely rewritten.

Signed-off-by: Vadim Zeitlin <vadim@zeitlins.org>

----

Notes:

 1. I couldn't make this work with the mix-ins as the original PR tried to do, they are just too confusing/magic (in my defence, the original patch didn't work at all though and mine at least does work).
 1. More generally speaking, the only redeemable quality of this patch is that it works with a small amount of changes. Ideally the email generation code should be rewritten to allow having multipart messages with text and HTML parts instead. But I won't have time to do this.
 1. Similarly, diff hunks should be parsed, but I didn't find any ready to use module to do it (I hoped cdiff could be used, but it's a program, not a library) and didn't have time to do this properly myself neither.
 1. I thought about adding a test for this but couldn't untangle how the test suite worked. I thought anything TAP-based would be simple to modify, but I was wrong.
 1. I didn't find any changelog-like file to modify, so I couldn't document this change.

moy commented:
Hi, 

Thanks a lot for the PR. Probably the most wanted feature in git-multimail.

This really means tests however. Did you read t/README and t/generate-test-emails?

Answering from my phone, will be back to civilization mid August.

moy commented:
Ah, and there is a [CHANGES](https://github.com/git-multimail/git-multimail/blob/master/git-multimail/CHANGES) which I use for release notes, you can add an entry there.

vadz commented:
Oops, I knew I had seen a change log somewhere, no idea how have I missed it. Will update, thanks.

And maybe will try to gather enough courage to look at the tests again...

vadz commented:
I've addressed your comments but the tests still elude me. I think the best would be to run them twice, once in plain text format and another time with `commitEmailFormat = html`, but where could this be plugged into the existing maze of scripts and makefiles is just beyond me.

moy commented:
Your entry in CHANGES was incorrect: 1.2.0 is indeed the next release, not the current one. I've fixed this, and added a test (after a bunch of changes to try to simplify it). See cce22e6.

Thanks for your contribution!

mhagger commented:
Is this safe from XSS attacks? Suppose an attacker can push to the Git repository. He controls (at least partly):

* The commit message
* The contents of the diff
* The author and committer usernames and email addresses
* The branch name

Is there any way he can use these fields to do something nasty?

moy commented:
I don't think there's a XSS possible attack, since the code uses `cgi.escape()` and I think it uses it on any user input that is rendered as HTML.

Actually, even the intro template is escaped, which can be considered as a mis-feature: this prevents the user from writing HTML in the templates (I've opened a bug for that: #116).

Also, AFAIK, XSS attacks in emails are far less serious than on website: an attacker could already send you an email with malicious HTML and your mailer is supposed to protect against it. (But I agree that this should not be an excuse to let XSS get in git-multimail!)

mhagger commented:
:+1: Cool; thanks for the response.
